Added an Install script

primis opened the pull request:
Proposal as an alternative solution to #30. Pull request #35 is the other option.
Also, added a man page.

veswill3 commented:
I was about a third of the way finished rewriting my version when I noticed this. Yeah, this is much better.
One thing I was going to add that you may (or may not) want to is moving the install features into a Makefile. It would add some nice seperation and make it compatible with [bpkg.io](bpkg.io).

To save 5 min if you want to include it, here is what I had in the Makefile:

```make
INSTALL_DIR := /usr/local/bin
COMPLETIONS_DIR := $(shell pkg-config --variable=completionsdir bash-completion)

install:
	install notes $(INSTALL_DIR)
	cp notes.bash_completion $(COMPLETIONS_DIR)/notes
	mkdir -p ~/.config/notes
	cp config.example ~/.config/notes

uninstall:
	rm -f $(INSTALL_DIR)/notes
	rm -f $(COMPLETIONS_DIR)/notes
```

primis commented:
Makefile is a good idea if it adds bpkg.io support, but bash completion breaks install if it's not installed. I'll make a subtarget for it though.

primis commented:
After a series of typo fixes, the Makefile has been confirmed working.

pimterry commented on Makefile:
```diff
@@ -0,0 +1,20 @@
+	install -m755 -D notes $(PREFIX)/bin/notes
+	install -m777 -D config.example $(USERDIR)/.config/notes/config.example
+	install -D notes.1 $(PREFIX)/share/man/man1/notes.1
+	mandb 1>/dev/null
```
On osx, this line fails, as there's no mandb. Doesn't seem to be necessary though - `man notes` works fine after installation. Can we make this optional, and silently skip it if `mandb` doesn't exist?

pimterry replied:
We could do with more error handling here too, since this could easily silently go wrong. We should `set -o errexit`, for starters, and trap the error to display a nice message like, along the lines of:

> Oh no, notes failed to install! Please follow the manual install steps at github.com/pimterry/notes, and file a bug (if one doesn't already exist) with the issue you're seeing, so we can get it fixed for you.

This bit could be a separate issue actually, doesn't need to block this. Feel free to file one and do those changes later if you'd prefer.

pimterry replied:
Some success output generally is probably a good idea too actually, to explain what's just happened and get people started:

> Notes has been installed to $(PREFIX)/bin/notes. Config has been created in $(exampleconfigpath), which you can edit if you'd like to change the default settings.
> Get started now by running `notes open my-new-note`, or `notes help` for more info.

primis replied:
> man notes works fine after installation. Can we make this optional, and silently skip it if mandb doesn't exist?

~~We can probably get away without mandb then. Thoughts?~~
Made it a silent fail if non-existent. it speeds up recognition on fedora at least, so I'm gonna keep it for now.

pimterry commented on Makefile:
```diff
@@ -0,0 +1,20 @@
+		cp notes.bash_completion "$(BASH_COMPLETION_DIR)/notes";\
+	fi # Small test for bash completion support
+	install -m755 -D notes $(PREFIX)/bin/notes
+	install -m777 -D config.example $(USERDIR)/.config/notes/config.example
```
We should just put the config in the right place by default I think, with no `.example`, but with the settings commented out so there's no effect unless you edit it. I think that's a fairly standard pattern.

primis replied:
I wholeheartedly agree. Fixed.

pimterry commented on README.md:
````diff
@@ -21,30 +21,13 @@ This is just one tool in the chain. `notes` is a command line tool, and some peo
 
 ```bash
-curl https://cdn.rawgit.com/pimterry/notes/v0.2.0/notes > /usr/local/bin/notes && chmod +x /usr/local/bin/notes
+curl https://cdn.rawgit.com/pimterry/notes/master/install.sh | bash
````
This probably won't work - that's a production rawgit url, which is cached permanently by URL (see rawgit.com), so it'll never pick up new versions. https://rawgit.com (no `cdn.`) should work.

primis replied:
Fixed

pimterry commented on README.md:
```diff
@@ -21,30 +21,13 @@ This is just one tool in the chain. `notes` is a command line tool, and some peo
 
 ## How do I install this?
 
-Download `notes`, `chmod +x`, put it in your `$path`. This will probably do it:
```
I don't think we should get rid of the existing instructions, since there'll be people with fiddly cases. We should split out an 'Easy install' section up front though, and 'Manual install' section for the fiddly cases.

pimterry replied:
This goes double for the bash completion section, which is still very relevant to anybody on OSX or probably Windows, as well as anybody with other non-standard setups.

primis replied:
Good point, I put them back

pimterry commented on install.sh:
```diff
@@ -0,0 +1,33 @@
+
+# Variable Definitions go here. 
+user_home=`eval echo ~$SUDO_USER`
+bash_completion_dir=`pkg-config --variable=completionsdir bash-completion 2>/dev/null`
```
This isn't used in this file is it? We could drop it.

primis replied:
Done :)

pimterry commented on Makefile:
```diff
@@ -0,0 +1,20 @@
+	@if [ -d $(BASH_COMPLETION_DIR) ]; then \
+	 	echo $(BASH_COMPLETION_DIR);\
+		cp notes.bash_completion "$(BASH_COMPLETION_DIR)/notes";\
+	fi # Small test for bash completion support
```
If we skip bash completion, we should print something to say so, and link to the instructions to add it by hand.

primis replied:
Added. Good idea.

pimterry commented on install.sh:
```diff
@@ -0,0 +1,33 @@
+
+# Make sure we have everything actually installed that we need to install this.
+echo "Checking for Dependencies..."
+    assertInstalled bash curl tar mktemp install make
```
I think this (and the root check too probably) should live in the makefile instead.

The `install` assert definitely, because that's only used in the Makefile, and what we need installed in there might change. It'd be good to have things specified in the same file they're used, to ensure they're consistent and make it easy to remember to keep them up to date. The root check is less important, but this script doesn't *require* root afaict, and if we can move as much as possible into the Makefile to keep things simpler rather than spread over two files that'd be great.

That would make this script responsible for fetching and starting the Makefile, and getting the minimum required for that to happen, and the Makefile responsible for all the rest.

primis replied:
I'm not sure if we could do that check for install inside Makefile, since making functions inside Makefiles is super hard. and I think that having one assertInstalled is better than two.

As a side note, Root is checked inside install.sh because it allows us to correctly identify the user directory, which would also be difficult if we had to elevate inside makefile.

pimterry replied:
Sounds like this is more effort then it's worth then - let's shelve it for now.

pimterry commented on Makefile:
```diff
@@ -0,0 +1,20 @@
+	install -m777 -D config.example $(USERDIR)/.config/notes/config.example
+	install -D notes.1 $(PREFIX)/share/man/man1/notes.1
+	mandb 1>/dev/null
+uninstall:
```
This is currently undocumented. We should explain how to uninstall too, both in the output of this script and in the readme.

primis replied:
I documented it in a novel way. Let me know if you approve

primis commented on install.sh:
```diff
@@ -0,0 +1,33 @@
+
+# Variable Definitions go here. 
+user_home=`eval echo ~$SUDO_USER`
+bash_completion_dir=`pkg-config --variable=completionsdir bash-completion 2>/dev/null`
```
Well, we check bash_completions here, then use it in the makefile. I could most likely move this into Makefile

primis commented:
I got through most of your suggestions, the only major one I'm holding onto is the dependency check / root elevation, as I'm incapable personally of figuring out how to do that in a makefile. Hope it gets to you well.

pimterry commented on Makefile:
```diff
@@ -0,0 +1,34 @@
+	@install -m755 -D notes $(PREFIX)/bin/notes
+	@install -m777 -D config $(USERDIR)/.config/notes/config
+	@install -D notes.1 $(PREFIX)/share/man/man1/notes.1
+	@mandb 1>/dev/nulli 2>&1 # Fail silently if we don't have a mandb
```
Typo: nulli -> null

primis replied:
That's what I get for using vim 😺

pimterry commented on install.sh:
```diff
@@ -0,0 +1,37 @@
+
+# Yay, Echo self documents! :D
+echo "Checking for root..."
+    [ "$(whoami)" != "root" ] && exec sudo -- "$0" "$@"
```
`sudo` doesn't (typically) preserve environmental variables, which means with this script you currently can't uninstall unless you're already running as root. If you're not, then `export uninstall=1; ./install.sh` will elevate itself, and then run a normal install.

primis replied:
Fixed this by moving uninstall to a argument instead of variable

pimterry commented on Makefile:
```diff
@@ -0,0 +1,34 @@
+	@install -m777 -D config $(USERDIR)/.config/notes/config
+	@install -D notes.1 $(PREFIX)/share/man/man1/notes.1
+	@mandb 1>/dev/nulli 2>&1 # Fail silently if we don't have a mandb
+	@echo -e "Notes has been installed to \033[0;32m$(PREFIX)/bin/notes." \
```
`echo -e` isn't supported in all shells, and I think generally [isn't recommended](http://askubuntu.com/a/537987/116547). We should probably move all of this to printf really, but I'd be fine with a working simple `echo` solution. In zsh on OSX this prefixes this message with `-e ` for me right now, which isn't great.

pimterry commented on README.md:
````diff
@@ -45,8 +52,16 @@ curl https://cdn.rawgit.com/pimterry/notes/v0.2.0/notes.bash_completion > /usr/s
+## What if I want to uninstall this?
+If you used the automated install script to install notes, you can uninstall it the same way by running:
+```bash
+export uninstall=1 && curl https://rawgit.com/pimterry/notes/master/install.sh | bash
````
Because of the sudo env vars thing, and because this is a little unorthodox, I think we should probably simplify, and just pass `uninstall` as a command line argument instead. `... | bash -s uninstall` should work fine for piping, and being able to manually pull it down and then run just `./install.sh uninstall` feels nice and clear.

pimterry commented:
Hi @primis - it's been a while, but I'm still pretty keen to get this in. Interested in finishing up the last few points above?

I have also just merged the zsh autocompletion PR in the meantime. We don't need to immediately include that in the setup for this (although feel free to have a go), but there will be a conflict in the readme. Should be a simple update, just need to make sure the zsh docs stay in.

primis commented:
@pimterry All changes have been implemented except for the shelved dependency check. I removed the echo -e statements and replaced them with simpler, non-colorized output too.

pimterry commented on Makefile:
```diff
@@ -0,0 +1,33 @@
+PREFIX ?= /usr/local
+BASH_COMPLETION_DIR := $(shell pkg-config --variable=completionsdir bash-completion) 
```
Not sure why I didn't spot this before, sorry. If I don't have `bash-completion` installed and findable with `pkg-config` (e.g. on OSX) then this all fails here with:

```
Package bash-completion was not found in the pkg-config search path.
Perhaps you should add the directory containing `bash-completion.pc'
to the PKG_CONFIG_PATH environment variable
No package 'bash-completion' found
make: *** [install] Error 127
```

pimterry replied:
I'm sure I tested this before, on this OSX laptop, no idea why it wasn't breaking then...

primis replied:
I don't have modern OS X, but this doesn't fail on my snow leopard server (which doesn't have bash-completion, but does have pkg-config), so I can't reproduce. Any ideas?

pimterry commented on install.sh:
```diff
@@ -0,0 +1,37 @@
+
+# This is an install script for notes
+
+function assertInstalled() {
```
Interesting minor but bizarre problem here. If I curl to install this with:

```
curl -L https://rawgit.com/pimterry/notes/install-test/install.sh | bash
```

(literally exactly that - I've pushed this PR + a tweak to make the URL work before this is merged to the `install-test` branch here so I can curl it for testing)

then I get:

```
Checking for root...
Checking for Dependencies...
bash: line 4: assertInstalled: command not found
Downloading and Extracting Notes from Repository...
```

Note the (ineffectual) `assertInstalled` error. If I run `./install.sh` from bash within a checkout of the same code though, I get the same result, and otherwise the same output, but without that error. Any idea why?

primis replied:
This can be reproduced with `cat install.sh | bash` and `cat install.sh | sh` as well, which means there's something fundamentally wrong with how I'm making a function.

primis replied:
Options here are to remove assertInstalled, and just assume we have everything, other option is to de-functionize it and put it inline. Thoughts?

primis replied:
Figured it out. It's because we're doing a switch to root, and the function is being curled as regular user. if you do that curl under root, it doesn't happen.

primis replied:
I fixed this one by moving root elevation *above* the function declaration.

pimterry commented on README.md:
````diff
@@ -56,7 +63,15 @@ curl https://cdn.rawgit.com/pimterry/notes/v0.2.0/notes._notes > /usr/share/zsh/
+## What if I want to uninstall this?
+If you used the automated install script to install notes, you can uninstall it the same way by running:
+```bash
+curl https://rawgit.com/pimterry/notes/master/install.sh | bash -s uninstall
````
Both this and the other curl above need to be `curl -L`. I think this is a new thing, but `rawgit.com` now includes a redirect, which curl doesn't follow by default.

primis commented:
I'm pretty confident the Makefile changes will work (please verify). But the other two issues are 100% taken care of.

primis commented:
Any Progress towards merge? 
-- Primis

pimterry commented:
@primis I hit a few OSX bugs in the installation - `echo -e` doesn't quite work right, the makefile didn't handle `mandb` not being present correctly - and there were some outstanding conflicts with master. I've gone ahead and just found and tested and pushed some fixes for the OSX issues straight onto your branch, hope that's ok.

I've also updated all the URLs (from this and the ones the readme too) to use `latest-release`, rather than using master or hardcoding a version, and we've now got a `latest-release` branch which I'll move around to just point to each released version as they're ready :smile:.

I think this all works for me now, so I'm ready to merge it. Can you test it still works ok for you? I've pushed a version with non-master URLs to `install-test` on this repo, so you should be able to check this with:

```
curl -L https://rawgit.com/pimterry/notes/install-test/install.sh | bash
```

and

```bash
curl -L https://rawgit.com/pimterry/notes/install-test/install.sh | bash -s -- uninstall
```

Does that do everything correctly for you?

primis commented:
Just tried it on fedora, it still works. Feel free to push.

On May 7, 2017, 11:09 AM -0400, Tim Perry <notifications@github.com>, wrote:
>
> @pimterry approved this pull request.
>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub (https://github.com/pimterry/notes/pull/36#pullrequestreview-36666746), or mute the thread (https://github.com/notifications/unsubscribe-auth/ABflg5d3vv39lNiS6qRSxSaebOvWo5GGks5r3d7FgaJpZM4LYp1G).
>
>
>

pimterry commented:
Perfect, thanks! Merged 🍾 🎆 👍
